File: frontend/src/api/werkzeugName.ts
// Werkzeug-Anzeigename + Typ-Labels (D34a, Issue #33).
//
// Spiegelt die Backend-Logik aus `camwosa/db/werkzeug_name.py` fuer eine
// sofortige Live-Vorschau im Editor (ohne API-Roundtrip pro Tastendruck).
// Quelle der Wahrheit beim Speichern bleibt das Backend (`_anzeigename`).

import type { Werkzeug, WerkzeugTyp } from "./types";

/** Anzeige-Labels je Werkzeug-Typ (echte Umlaute). */
export const WERKZEUG_TYP_LABEL: Record<WerkzeugTyp, string> = {
  schaftfraeser: "Schaftfräser",
  kugelfraeser: "Kugelfräser",
  torusfraeser: "Torusfräser",
  v_bit: "V-Bit",
  ballnose_v_bit: "Ballnose-V-Bit",
  gravierstichel: "Gravierstichel",
  bohrer: "Bohrer",
  einschneider: "Einschneider",
  fischschwanz: "Fischschwanz",
  schruppfraeser: "Schruppfräser",
  diamantgravierer: "Diamantgravierer",
  drag_gravierer: "Schleppgravierer",
};

/** Konische Typen zeigen den Spitzenwinkel statt der Schneidenzahl. */
const KONISCH: ReadonlySet<WerkzeugTyp> = new Set<WerkzeugTyp>([
  "v_bit", "ballnose_v_bit", "gravierstichel",
  "diamantgravierer", "drag_gravierer",
]);

export function istKonisch(typ: WerkzeugTyp): boolean {
  return KONISCH.has(typ);
}

/** Durchmesser kompakt: 6 statt 6.0, 12.7 bleibt 12.7. */
function fmtMm(wert: number): string {
  if (Math.abs(wert - Math.round(wert)) < 1e-6) return String(Math.round(wert));
  return wert.toFixed(2).replace(/0+$/, "").replace(/\.$/, "");
}

/** Auto-Name aus den Werkzeug-Daten (ohne Zusatz). */
export function werkzeugAutoName(w: Partial<Werkzeug>): string {
  const typ = (w.typ ?? "schaftfraeser") as WerkzeugTyp;
  const teile: string[] = [WERKZEUG_TYP_LABEL[typ] ?? typ];

  if (KONISCH.has(typ) && w.spitzenwinkel != null) {
    teile.push(`${fmtMm(w.spitzenwinkel)}°`);
  }
  if (w.durchmesser != null) {
    teile.push(`Ø${fmtMm(w.durchmesser)} mm`);
  }
  if (!KONISCH.has(typ) && w.schneiden) {
    teile.push(`${w.schneiden}-Schneider`);
  }
  if (w.material) {
    teile.push(String(w.material));
  }
  return teile.join(" · ");
}

/** Voller Anzeigename: Auto-Name + optionaler Zusatz. */
export function werkzeugAnzeigename(w: Partial<Werkzeug>): string {
  const basis = werkzeugAutoName(w);
  const zusatz = (w.name_zusatz ?? "").trim();
  return zusatz ? `${basis} (${zusatz})` : basis;
}

/** Bevorzugt den Backend-Anzeigenamen, faellt sonst auf lokale Berechnung zurueck. */
export function anzeigename(w: Werkzeug): string {
  return w._anzeigename || werkzeugAnzeigename(w) || w.name || w.id;
}

File: backend/camwosa/db/werkzeug_name.py
"""Werkzeug-Anzeigename automatisch aus den Daten (Cluster D34a, Issue #33).

Markus' Anforderung: der Werkzeug-Name soll sich immer aus den Daten ergeben
(Typ + Durchmesser + Schneiden + Material + ggf. Winkel), und am Ende darf ein
optionaler eigener Zusatz (`name_zusatz`) angehängt werden.

Beispiele:
    Schaftfräser Ø6 mm · 2-Schneider · Hartmetall
    V-Bit 60° Ø12.7 mm · Hartmetall
    Kugelfräser Ø3 mm · 2-Schneider · Hartmetall (mein Liebling)
"""

from __future__ import annotations

from camwosa.db.models import Werkzeug, WerkzeugTyp

_TYP_LABEL: dict[WerkzeugTyp, str] = {
    WerkzeugTyp.SCHAFTFRAESER: "Schaftfräser",
    WerkzeugTyp.KUGELFRAESER: "Kugelfräser",
    WerkzeugTyp.TORUSFRAESER: "Torusfräser",
    WerkzeugTyp.V_BIT: "V-Bit",
    WerkzeugTyp.BALLNOSE_V_BIT: "Ballnose-V-Bit",
    WerkzeugTyp.GRAVIERSTICHEL: "Gravierstichel",
    WerkzeugTyp.BOHRER: "Bohrer",
    WerkzeugTyp.EINSCHNEIDER: "Einschneider",
    WerkzeugTyp.FISCHSCHWANZ: "Fischschwanz",
    WerkzeugTyp.SCHRUPPFRAESER: "Schruppfräser",
    WerkzeugTyp.DIAMANTGRAVIERER: "Diamantgravierer",
    WerkzeugTyp.DRAG_GRAVIERER: "Schleppgravierer",
}

# Konische Typen zeigen den Spitzenwinkel statt der Schneidenzahl prominent.
_KONISCH = {
    WerkzeugTyp.V_BIT, WerkzeugTyp.BALLNOSE_V_BIT,
    WerkzeugTyp.GRAVIERSTICHEL, WerkzeugTyp.DIAMANTGRAVIERER,
    WerkzeugTyp.DRAG_GRAVIERER,
}


def _fmt_mm(wert: float) -> str:
    """Durchmesser kompakt: 6 statt 6.0, 12.7 bleibt 12.7."""
    if abs(wert - round(wert)) < 1e-6:
        return str(int(round(wert)))
    return f"{wert:.2f}".rstrip("0").rstrip(".")


def werkzeug_auto_name(werkzeug: Werkzeug) -> str:
    """Generiert den Anzeigenamen aus den Werkzeug-Daten (ohne Zusatz)."""
    typ_label = _TYP_LABEL.get(werkzeug.typ, werkzeug.typ.value)
    teile: list[str] = [typ_label]

    if werkzeug.typ in _KONISCH and werkzeug.spitzenwinkel:
        teile.append(f"{_fmt_mm(werkzeug.spitzenwinkel)}°")

    teile.append(f"Ø{_fmt_mm(werkzeug.durchmesser)} mm")

    # Schneidenzahl (bei nicht-konischen Fräsern aussagekräftig)
    if werkzeug.typ not in _KONISCH and werkzeug.schneiden:
        teile.append(f"{werkzeug.schneiden}-Schneider")

    # Material (Hartmetall/HSS/…) — nur wenn aussagekräftig
    if werkzeug.material:
        teile.append(werkzeug.material.value)

    return " · ".join(teile)


def werkzeug_anzeigename(werkzeug: Werkzeug) -> str:
    """Voller Anzeigename: Auto-Name + optionaler eigener Zusatz."""
    basis = werkzeug_auto_name(werkzeug)
    zusatz = (werkzeug.name_zusatz or "").strip()
    if zusatz:
        return f"{basis} ({zusatz})"
    return basis


__all__ = ["werkzeug_anzeigename", "werkzeug_auto_name"]
